src/iceberg/transform.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/transform.h
+
+#include <cstdint>
+#include <memory>
+
+#include "iceberg/arrow_c_data.h"
+#include "iceberg/error.h"
+#include "iceberg/expected.h"
+#include "iceberg/iceberg_export.h"
+#include "iceberg/type_fwd.h"
+#include "iceberg/util/formattable.h"
+
+namespace iceberg {
+
+/// \brief Transform types used for partitioning
+enum class TransformType : uint8_t {
+  /// Used to represent some customized transform that can't be recognized or supported
+  /// now.
+  kUnknown,
+  /// Equal to source value, unmodified
+  kIdentity,
+  /// Hash of value, mod `N`
+  kBucket,
+  /// Value truncated to width `W`
+  kTruncate,
+  /// Extract a date or timestamp year, as years from 1970
+  kYear,
+  /// Extract a date or timestamp month, as months from 1970-01-01
+  kMonth,
+  /// Extract a date or timestamp day, as days from 1970-01-01
+  kDay,
+  /// Extract a timestamp hour, as hours from 1970-01-01 00:00:00
+  kHour,
+  /// Always produces `null`
+  kVoid,
+};
+
+/// \brief Get the relative transform name
+constexpr std::string_view ToString(TransformType type) {
+  /// Transform names used for formatting
+  constexpr std::string_view TransformTypeNames[] = {"Unknown",  "Identity", "Bucket",
+                                                     "Truncate", "Year",     "Month",
+                                                     "Day",      "Hour",     "Void"};
+  return TransformTypeNames[static_cast<size_t>(type)];
+}
+
+/// \brief A transform function used for partitioning.
+class ICEBERG_EXPORT TransformFunction : public util::Formattable {
+ public:
+  explicit TransformFunction(TransformType type);
+  /// \brief Transform an input array to a new array
+  virtual expected<ArrowArray, Error> Transform(const ArrowArray& inputArrowArray) = 0;
```

Comment:
We use snake case for parameter name. So either `input_arrow_array` or simply `data`?